<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-базы-данных" data-toc-modified-id="Подключение-базы-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение базы данных</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Анализ" data-toc-modified-id="Анализ-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ</a></span><ul class="toc-item"><li><span><a href="#Количестов-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количестов-книг,-вышедших-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Количестов книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Средняя-оценка-и-количество-обзоров" data-toc-modified-id="Средняя-оценка-и-количество-обзоров-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Средняя оценка и количество обзоров</a></span></li><li><span><a href="#Самое-крупное-издательство-книнг" data-toc-modified-id="Самое-крупное-издательство-книнг-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Самое крупное издательство книнг</a></span></li><li><span><a href="#Самый-высокорейтинговый-автор" data-toc-modified-id="Самый-высокорейтинговый-автор-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Самый высокорейтинговый автор</a></span></li><li><span><a href="#Среднее-количество-обзоров-пользователей" data-toc-modified-id="Среднее-количество-обзоров-пользователей-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Среднее количество обзоров пользователей</a></span></li></ul></li><li><span><a href="#Результаты" data-toc-modified-id="Результаты-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Результаты</a></span></li></ul></div>

# Анализ данных базы данных об онлайн-книгах.

Предварительный анализ базы данных с информацией о книгах, издательствах, авторах.

## Подключение базы данных

In [1]:
!pip install sqlalchemy 

In [2]:
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con=engine.connect()

## Описание данных

Схема базы данных

![Альтернативный текст для людей с ограниченными возможностями зрения](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

### Таблица books

Выведем первые строки таблицы

In [3]:
# чтобы выполнить SQL-запрос, используем Pand
query = '''SELECT * FROM books LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


Таблица `books` содержит данные о книгах:
* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

### Таблица authors

Выведем первые строки таблицы

In [4]:
query = '''SELECT * FROM authors LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


Таблица `authors` содержит данные об авторах:
* `author_id` — идентификатор автора;
* `author` — имя автора.

### Таблица publishers

Выведем первые строки таблицы

In [5]:
query = '''SELECT * FROM publishers LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


Таблица `publishers` содержит данные об издательствах:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

### Таблица ratings

Выведем первые строки таблицы

In [6]:
query = '''SELECT * FROM ratings LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


Таблица `ratings` данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

### Таблица reviews

Выведем первые строки таблицы

In [7]:
query = '''SELECT * FROM reviews LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


Таблица `reviews` данные о пользовательских обзорах:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.

## Анализ

###  Количестов книг, вышедших после 1 января 2000 года

Посчитайтаем, сколько книг вышло после 1 января 2000 года:

In [8]:
query = '''SELECT  COUNT(*)
FROM books 
   WHERE publication_date>'2000-01-01' '''
pd.io.sql.read_sql(sql=text(query), con = con)

,count
0,819


После 1 января 2000 года вышло 819 книг.

### Средняя оценка и количество обзоров

Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [9]:

query = '''
SELECT rv.book_id, rv.title,  rv.cnt, rt.avg_rate FROM
    (SELECT b.book_id, b.title, COUNT(r.review_id) AS cnt 
        FROM books b
        LEFT JOIN reviews r ON b.book_id = r.book_id
        GROUP BY b.book_id,b.title
        
    ) AS rv
    JOIN 
    (SELECT b.book_id, AVG(r.rating) AS avg_rate
        FROM books b
        LEFT JOIN ratings r ON b.book_id = r.book_id
        GROUP BY b.book_id
        ORDER BY  avg_rate
    ) AS rt ON rv.book_id = rt.book_id '''
pd.io.sql.read_sql(sql=text(query), con = con)

,book_id,title,cnt,avg_rate
0,652,The Body in the Library (Miss Marple #3),2,4.500000
1,273,Galápagos,2,4.500000
2,51,A Tree Grows in Brooklyn,5,4.250000
3,951,Undaunted Courage: The Pioneering First Missio...,2,4.000000
4,839,The Prophet,4,4.285714
...,...,...,...,...
995,64,Alice in Wonderland,4,4.230769
996,55,A Woman of Substance (Emma Harte Saga #1),2,5.000000
997,148,Christine,3,3.428571
998,790,The Magicians' Guild (Black Magician Trilogy #1),2,3.500000


Средняя оценка книг находится в диапазоне от 3 до 4.5. Количество обзоров от 0 до 7.

### Самое крупное издательство книнг

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.

In [10]:
query = '''
SELECT p.publisher, COUNT(*) 
    FROM publishers  p
    JOIN books b ON p.publisher_id = b.publisher_id
    WHERE b.num_pages>50
    GROUP BY p.publisher_id, p.publisher
    ORDER BY COUNT(*) DESC
    LIMIT 1'''
pd.io.sql.read_sql(sql=text(query), con = con)

,publisher,count
0,Penguin Books,42


Самое крупное книжное издательство — "Penguin Books". Оно выпустило 42 книги, объемом более 50 страниц

### Самый высокорейтинговый автор

Определим автора с самой высокой средней оценкой книг. 

In [11]:
query = '''
SELECT a.author, AVG(bs.rate)
    FROM authors a
    JOIN 
    (SELECT b.book_id, b.author_id, AVG(r.rating) as rate
        FROM books b
        JOIN ratings r ON b.book_id = r.book_id
        GROUP BY b.book_id, b.author_id
        HAVING COUNT(*)>=50
    ) AS bs ON a.author_id = bs.author_id
    GROUP BY a.author_id, a.author
    ORDER BY AVG(bs.rate) DESC
    LIMIT 1'''
pd.io.sql.read_sql(sql=text(query), con = con)

,author,avg
0,J.K. Rowling/Mary GrandPré,4.283844


Автор с самыми высокой средней оценкой книг — GrandPré. Средняя оценка 4.3.

### Среднее количество обзоров пользователей

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [12]:
query = '''
SELECT AVG(rv.cnt) 
    FROM
    (SELECT COUNT(*) as cnt
        FROM reviews rv
        JOIN
        (SELECT username
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 48
        ) u on rv.username = u.username
         GROUP BY u.username
    ) AS rv'''
pd.io.sql.read_sql(sql=text(query), con = con)

,avg
0,24.0


Среднее количество обзоров у активных пользовтаелей — 24.

## Результаты


* После 1 января 2000 года вышло 819 книг.
* Средняя оценка книг находится в диапазоне от 3 до 4.5. Количество обзоров от 0 до 7.
* Книги с самыми высокими оценками -- "The Body in the Library (Miss Marple #3)" и "Galápagos"
* Самое крупное книжное издательство — "Penguin Books". Оно выпустило 42 книги, объемом более 50 страниц
* Автор с самыми высокой средней оценкой книг — GrandPré. Средняя оценка 4.3.
* Среднее количество обзоров у активных пользовтаелей — 24.

* Работа с ассортиментом
    * Популярные книги: Выделить книги с самыми высокими средними оценками ("The Body in the Library (Miss Marple #3)", "Galápagos" и другие) в категории "Рекомендованные" или "Лучшие по оценкам"
    * Издательства: Плотнее сотрудничать с Penguin Books, так как у них наибольшее количество книг . Возможно, они могут предоставить эксклюзивные предложения.
    * Автор с высокой оценкой: Выделить книги GrandPré, так как у него высокая средняя оценка (4.3). Возможно, добавить его в подборки "Лучшие авторы" или предложить скидки на его книги.

* Маркетинг и продвижение
    * Продвижение книг с рейтингом 3–4.5: Раз рейтинг большинства книг в этом диапазоне, можно создать подборки вроде "Скрытые жемчужины" или "Незаслуженно забытые книги"
    * Фильтрация каталога: Добавить фильтр по средней оценке, чтобы пользователи могли искать книги с наивысшими рейтингами.
    * Тематика "Penguin Books": Если издательство активно выпускает книги в определенных жанрах (например, классика, научпоп), можно делать подборки книг этого издательства.

* Анализ и стратегия роста
    * Анализ жанров: Определить, какие жанры наиболее популярны среди книг с высокими оценками.
    * Рост базы отзывов: Если мало книг с большим числом рецензий, важно стимулировать обсуждение (форумы, email-рассылки с напоминанием оставить отзыв).
    * Изучение новых авторов: Проверить, какие авторы, кроме GrandPré, имеют высокие оценки, и усилить их продвижение.

Таким образом, главные направления работы магазина: улучшение отзывов, работа с популярными авторами и издательствами, маркетинговые подборки и фильтрация контента по оценкам.